### PR TITLE
Create resources in proper namespaces

### PR DESCRIPTION
**Pull request by sacharya**
Major changes:
- Check ClusterDeployment exists for the clusterName
- Create BareMetalHosts in the managed cluster in the "openshift-machine-api" namespace. Fixes [#15](https://github.com/mhrivnak/multicluster-inventory/issues/15)
- Create Hive SyncSet in the hub cluster in the same namespace as ClusterDeployment.
- Change the operator to cluster-scoped, so the operator can run in the operator-namespace and manage syncsets in the clusterdeployment namespace.

### CODE REVIEW

File: deploy/role_binding.yaml [hunk 1]
```diff
@@ -1,11 +1,12 @@
-kind: RoleBinding
+kind: ClusterRoleBinding
 apiVersion: rbac.authorization.k8s.io/v1
 metadata:
   name: multicluster-inventory
 subjects:
 - kind: ServiceAccount
   name: multicluster-inventory
+  namespace: default
```
Comment (sacharya) [inline deploy/role_binding.yaml]: What namespace should the operator run on? Put default for now.

File: pkg/apis/midas/v1alpha1/baremetalasset_types.go [hunk 2]
```diff
@@ -6,6 +6,9 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// ManagedClusterResourceNamespace is the namespace on the managed cluster where BareMetalHosts are placed.
+const ManagedClusterResourceNamespace string = "openshift-machine-api"
```
Comment (sacharya) [inline pkg/apis/midas/v1alpha1/baremetalasset_types.go]: If the namespace on the managed cluster can change, we can make it configurable via BMA spec.

**Comment by djzager** [comment]
Should update the description to mention that it fixes #15

File: pkg/controller/baremetalasset/baremetalasset_controller.go [hunk 3]
```diff
@@ -189,28 +198,89 @@ func (r *ReconcileBareMetalAsset) Reconcile(request reconcile.Request) (reconcil
...
+		}
+
+		// If clusterDeployment is found, ensure syncset is created
+		err = r.ensureHiveSyncSet(instance, foundCd, reqLogger)
 		if err != nil {
 			return reconcile.Result{}, err
 		}
 	} else {
-		// if clusterName is not specified, delete the syncset if it exists
+		// Without a clusterName, we do not know what namespace the syncset is in.
```
Comment (djzager) [inline pkg/controller/baremetalasset/baremetalasset_controller.go]: I think working off the assumption that the ClusterDeployment exists in the same namespace as the BareMetalAssets makes this step unnecessary.

File: pkg/controller/baremetalasset/baremetalasset_controller.go [hunk 4]
```diff
@@ -219,8 +289,8 @@ func (r *ReconcileBareMetalAsset) Reconcile(request reconcile.Request) (reconcil
 	return reconcile.Result{}, nil
 }
 
-func (r *ReconcileBareMetalAsset) ensureHiveSyncSet(bma *midasv1alpha1.BareMetalAsset, reqLogger logr.Logger) error {
-	hsc := r.newHiveSyncSet(bma, reqLogger)
+func (r *ReconcileBareMetalAsset) ensureHiveSyncSet(bma *midasv1alpha1.BareMetalAsset, cd *hivev1.ClusterDeployment, reqLogger logr.Logger) error {
```
Comment (djzager) [inline pkg/controller/baremetalasset/baremetalasset_controller.go]: It doesn't look like these changes are necessary if we assume that BareMetalAssets exist in the same namespace as the ClusterDeployment.

File: pkg/controller/baremetalasset/baremetalasset_controller.go [hunk 5]
```diff
@@ -302,7 +372,7 @@ func (r *ReconcileBareMetalAsset) ensureHiveSyncSet(bma *midasv1alpha1.BareMetal
 	return r.client.Status().Update(context.TODO(), bma)
 }
 
-func (r *ReconcileBareMetalAsset) newHiveSyncSet(bma *midasv1alpha1.BareMetalAsset, reqLogger logr.Logger) *hivev1.SyncSet {
+func (r *ReconcileBareMetalAsset) newHiveSyncSet(bma *midasv1alpha1.BareMetalAsset, cd *hivev1.ClusterDeployment, reqLogger logr.Logger) *hivev1.SyncSet {
```
Comment (djzager) [inline pkg/controller/baremetalasset/baremetalasset_controller.go]: It doesn't look like these changes are necessary if we assume that BareMetalAssets exist in the same namespace as the ClusterDeployment.

File: pkg/controller/baremetalasset/baremetalasset_controller.go [hunk 6]
```diff
@@ -80,6 +80,15 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
+	// Watch for changes to secondary resource ClusterDeployments and requeue the owner BareMetalAsset
+	err = c.Watch(&source.Kind{Type: &hivev1.ClusterDeployment{}}, &handler.EnqueueRequestForOwner{
```
[the remark this answers is not in the document]
Comment (djzager) [reply pkg/controller/baremetalasset/baremetalasset_controller.go]:
```suggestion
	err = c.Watch(&source.Kind{Type: &hivev1.ClusterDeployment{}}, &handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(clusterDeploymentHandlerFunc)})
```

then you can define a new function `clusterDeploymentHandlerFunc` that can enqueue events on ClusterDeployments -> requests for **all** BareMetalAssets that are associated with that ClusterDeployment. Something like:

```
func (r *ReconcileBareMetalAsset) clusterDeploymentHandlerFunc(a handler.MapObject) []reconcile.Request {
  clusterDeployment, ok := a.Object.(*hivev1.ClusterDeployment)
  // handle err

  // get all BareMetalAssets that belong to this ClusterDeployment
  // not immediately clear how we do that...
  // get all BareMetalAssets and look at them one by one (sounds expensive)
  // or maybe use labelselector

  // create a slice of requests
  requests := []reconcile.Request{}

  // for each BareMetalAsset that belongs to this ClusterDeployment
  requests = append(requests, reconcile.Request{
    NamespacedName: types.NamespacedName{
      Name: bma.Name,
      Namespace: bma.Namespace,
    },
  })

  return requests
}
```

[the remark this answers is not in the document]
Comment (sacharya) [reply pkg/controller/baremetalasset/baremetalasset_controller.go]: Thanks for the tip. This means we don't have to requeue the request when Secret and ClusterDeployment are not found, and have it keep trying. The watch will wake up the controller when they are found.

File: pkg/controller/baremetalasset/baremetalasset_controller.go [hunk 7]
```diff
@@ -189,28 +198,89 @@ func (r *ReconcileBareMetalAsset) Reconcile(request reconcile.Request) (reconcil
...
+					Status:  corev1.ConditionFalse,
+					Reason:  "ClusterDeploymentNotFound",
+					Message: fmt.Sprintf("A cluster deployment with the name %v in namespace %v could not be found", clusterName, clusterName),
+				})
+				statusErr := r.client.Status().Update(context.TODO(), instance)
+				if statusErr != nil {
+					reqLogger.Error(err, "Failed to update instance status")
+					return reconcile.Result{}, statusErr
+				}
+				return reconcile.Result{Requeue: true}, nil
```
[the remark this answers is not in the document]
Comment (sacharya) [reply pkg/controller/baremetalasset/baremetalasset_controller.go]: We don't need the requeue anymore with the watch implementation.

**Comment by sacharya** [comment]
Just to summarize our discussion, we are assuming BareMetalAssets can be created independently, before we know what cluster they are going to be associated with. So they will live in their own namespace.A cluster is provisioned and ClusterDeployment in created in a clusterNamespace. A BareMetalHost can then be associated with a cluster by updating the clusterName field in the spec.

File: pkg/controller/baremetalasset/baremetalasset_controller.go [hunk 8]
```diff
@@ -80,6 +80,40 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
+					requests = append(requests, reconcile.Request{
+						NamespacedName: types.NamespacedName{
+							Name:      bma.Name,
+							Namespace: bma.Namespace,
+						},
+					})
+				}
+				return requests
+			}),
+		})
```
Comment (djzager) [inline pkg/controller/baremetalasset/baremetalasset_controller.go]: total nit
```suggestion
		}
	)
```

File: pkg/controller/baremetalasset/baremetalasset_controller.go [hunk 9]
```diff
@@ -189,28 +222,80 @@ func (r *ReconcileBareMetalAsset) Reconcile(request reconcile.Request) (reconcil
...
+			}
+		}
+		if hscNamespace == "" {
+			// No syncset found in relatedObjects. Nothing to do.
+			return reconcile.Result{}, nil
+		}
+		// If clusterName is not specified, delete the syncset if it exists
 		found := &hivev1.SyncSet{}
-		err := r.client.Get(context.TODO(), types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace}, found)
+		err := r.client.Get(context.TODO(), types.NamespacedName{Name: instance.Name, Namespace: hscNamespace}, found)
```
Comment (djzager) [inline pkg/controller/baremetalasset/baremetalasset_controller.go]: Why not skip all of this extra stuff and just:

```
err = r.client.Delete(context.TODO(), &hivev1.SyncSet{
    ObjectMeta: metav1.ObjectMeta{
        Namespace: instance.Name,
        Name:      hscNamespace,
    },
})
if err != nil {
  if !errors.IsNotFound(err) {
    reqLogger.Error(err, "Failed to delete Hive SyncSet")
    return reconcile.Result{}, err
  }
  reqLogger.Info("Hive SyncSet not found")
}
// remove from related objects & update status.
```

Comment (sacharya) [reply pkg/controller/baremetalasset/baremetalasset_controller.go]: Doing a GET call so I can keep the access to the `found` instance of SyncSet, which I can then use to delete the object reference from instance status.

File: pkg/controller/baremetalasset/baremetalasset_controller.go (hunk 8 above)
Comment (sacharya) [reply pkg/controller/baremetalasset/baremetalasset_controller.go]: That's the way the compiler wants it. ;)

File: pkg/controller/baremetalasset/baremetalasset_controller.go (hunk 9 above)
Comment (djzager) [reply pkg/controller/baremetalasset/baremetalasset_controller.go]: You already have the objectReference when you are `for _, ro := range instance.Status.RelatedObjects {` just keep it and use that to inform the DELETE and the `RemoveObjectReference()`

Comment (sacharya) [reply pkg/controller/baremetalasset/baremetalasset_controller.go]: Sweet. Fixed.

**Pull request merged**